Implement Chain with Option fuses

cuviper opened the pull request:
The iterators are now "fused" with `Option` so we don't need separate state to track which part is already exhausted, and we may also get niche layout for `None`. We don't use the real `Fuse` adapter because its specialization for `FusedIterator` unconditionally descends into the iterator, and that could be expensive to keep revisiting stuff like nested chains. It also hurts compiler performance to add more iterator layers to `Chain`.

This change was inspired by the [proposal](https://internals.rust-lang.org/t/proposal-implement-iter-chain-using-fuse/12006) on the internals forum. This is an alternate to #70332, directly employing some of the same `Fuse` optimizations as #70366 and #70750.

r? @scottmcm

cuviper commented:
@bors try @rust-timer queue

Locally, I measure this being a little faster than nightly on `deeply-nested-opt`, but a little slower on `deeply-nested-debug`, which is probably OK. It also doesn't suffer from #70749.

@KrishnaSannasi I went ahead and tried your folding idea for `nth` too, but that added ~40% instruction count to `deeply-nested-opt`. We might still justify that if it helps the runtime performance of `Chain::nth` with internal iteration, but I'll let you explore that separately.

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 859b8da21fc82c08fde8b3aafefc12881a0c3f09 with merge 1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"859b8da21fc82c08fde8b3aafefc12881a0c3f09","merge_sha":"1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=25690)
Build commit: 1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a (`1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=25690"},"merge_sha":"1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a"} -->

rust-timer commented:
Queued 1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a with parent 42abbd8878d3b67238f3611b0587c704ba94f39c, future [comparison URL](https://perf.rust-lang.org/compare.html?start=42abbd8878d3b67238f3611b0587c704ba94f39c&end=1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a).

scottmcm commented on src/libcore/iter/adapters/chain.rs:
```diff
@@ -57,128 +53,99 @@ where
+        match self {
+            Chain { a: Some(a), b: Some(b) } => a.count() + b.count(),
+            Chain { a: Some(a), b: None } => a.count(),
+            Chain { a: None, b: Some(b) } => b.count(),
```
This may emit the code for count 4 times, instead of just 2, so consider doing this instead as `(match self.a { ... }) + (match self.b { ... })`.

scottmcm commented on src/libcore/iter/adapters/chain.rs:
```diff
@@ -57,128 +53,99 @@ where
-            ChainState::Front => self.a.last(),
-            ChainState::Back => self.b.last(),
+            Chain { a: Some(a), b: None } => a.last(),
+            Chain { a: None, b: Some(b) } => b.last(),
```
Similar comment as in `count` -- consider ensuring there's only one callsite for each.

rust-timer commented:
Finished benchmarking try commit 1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a, [comparison URL](https://perf.rust-lang.org/compare.html?start=42abbd8878d3b67238f3611b0587c704ba94f39c&end=1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a).

cuviper commented:
@scottmcm I updated `count` and `last` as you suggested -- locally it didn't seem to make much difference, for better or worse. I tried `size_hint` too, since that's a similar pattern, but that regressed performance quite a bit, so I dropped it. Perhaps you were wise not to ask for that one. :slightly_smiling_face:

cuviper commented:
That perf report matches my own results -- better on `deeply-nested-opt`, worse on `deeply-nested-debug`, and slight improvement / neutral on the rest.

We can also [compare to #70332](https://perf.rust-lang.org/compare.html?start=601c05538b7638fb0937bd659aab41fdfb41444d&end=1f6bcab4a99b6f10ce766a1fe581010d3dd68f6a), with the caveat that they're not on the exact same merge base, but this PR looks a lot better.

scottmcm commented:
@bors r+

> Perhaps you were wise not to ask for that one. 🙂

I almost did 🙂  But then I thought about it more and worried that the `saturating_add` is more scary to be doing every time than the wraps-in-release add in `count`, for example.

bors commented:
:pushpin: Commit ce8abc63a7e1fcac4b69574e00a70352e583cec8 has been approved by `scottmcm`

<!-- @bors r=scottmcm ce8abc63a7e1fcac4b69574e00a70352e583cec8 -->
<!-- homu: {"type":"Approved","sha":"ce8abc63a7e1fcac4b69574e00a70352e583cec8","approver":"scottmcm"} -->

pnkfelix commented:
This seems like it injected regression noted in issue #71359. 

The regression in question *might* be fairly niche.

But I still wanted to check in about whether the benefit from this PR is significant enough to justify the risk associated with issue #71359? 

cc @cuviper and @scottmcm

cuviper commented:
IMO that's niche enough to not worry about -- they were using a _very_ long chain, but even that is now resolved by <https://github.com/bodil/typed-html/pull/117> (though not published yet). Plus, that project is already rife with [increased `recursion_limit`](https://github.com/bodil/typed-html/search?q=recursion_limit&type=Code), as high as 1024 in one example.

Also, this PR had an incidental benefit on the very long compilation time in #70749 -- that code now compiles quickly with the `Option`-based `Chain`.
